activesupport/test/test_case_test.rb

```diff
@@ -350,6 +350,85 @@ def test_assert_no_changes_with_long_string_wont_output_everything
   end
 end
 
+class ExceptionsInsideAssertionsTest < ActiveSupport::TestCase
+  def before_setup
+    require "stringio"
+    @out = StringIO.new
+    self.tagged_logger = ActiveSupport::TaggedLogging.new(Logger.new(@out))
+    super
+  end
+
+  def test_warning_is_logged_if_caught_internally
+    run_test_that_should_pass_and_log_a_warning
+    expected = <<~MSG
+      ExceptionsInsideAssertionsTest - test_warning_is_logged_if_caught_internally: assert_nothing_raised called but ArgumentError raised.
+      Assertions nested *inside* `assert_raises` are ignored, this could cause tests to pass incorrectly or fail confusingly.
+      Ensure your tests still pass if you change the order of nested assertions!
+    MSG
+    assert @out.string.include?(expected)
+  end
+
+  def test_warning_is_not_logged_if_caught_correctly_by_user
+    run_test_that_should_pass_and_not_log_a_warning
+    assert_not @out.string.include?("assert_nothing_raised")
+  end
+
+  def test_warning_is_not_logged_if_assertions_are_nested_correctly
+    error = assert_raises(Minitest::Assertion) do
+      run_test_that_should_fail_but_not_log_a_warning
+    end
+    assert_not @out.string.include?("assert_nothing_raised")
+    assert error.message.include?("(lambda)> changed")
+  end
+
+  def test_fails_and_warning_is_logged_if_wrong_error_caught
+    error = assert_raises(Minitest::Assertion) do
+      run_test_that_should_fail_confusingly
+    end
+    expected = <<~MSG
+      ExceptionsInsideAssertionsTest - test_fails_and_warning_is_logged_if_wrong_error_caught: assert_nothing_raised called but ArgumentError raised.
+      Assertions nested *inside* `assert_raises` are ignored, this could cause tests to pass incorrectly or fail confusingly.
+      Ensure your tests still pass if you change the order of nested assertions!
+    MSG
+    assert @out.string.include?(expected)
+    assert error.message.include?("ArgumentError: ArgumentError")
+    assert error.message.include?("in `block (2 levels) in run_test_that_should_fail_confusingly'")
+  end
+
+  private
+    def run_test_that_should_pass_and_log_a_warning
+      assert_raises(Minitest::UnexpectedError) do # this assertion passes, but it's unlikely to be how anyone writes a test
+        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the error logged.
+          raise ArgumentError.new
+        end
+      end
+    end
+
+    def run_test_that_should_fail_confusingly
+      assert_raises(ArgumentError) do # this assertion fails (confusingly) because it catches a Minitest::UnexpectedError.
+        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the error logged.
```

Comment:
```suggestion
        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the warning logged.
          raise ArgumentError.new
        end
      end
    end

    def run_test_that_should_fail_confusingly
      assert_raises(ArgumentError) do # this assertion fails (confusingly) because it catches a Minitest::UnexpectedError.
        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the warning logged.
```